### ISSUE TITLE
publish --canary should have some way of limiting publish to changed packages only

### ISSUE THREAD

**Issue by darrylhodgins** [lerna/lerna#962]
## Expected Behavior
In much the same way that `lerna publish` behaves (by only publishing packages with changes, plus packages with dependencies on those), `lerna publish --canary` should have some way to restrict publishing to only those packages with changes.

## Current Behavior
Currently, publish --canary forces inclusion of all packages ([here, in UpdatedPackagesCollector.js](https://github.com/lerna/lerna/blob/950e26f503a955982e564f95dd99b1b89ed5abcd/src/UpdatedPackagesCollector.js#L160))

## Possible Solution
Use the `--since` flag to filter out unchanged packages.

## Context
In a project with a large number of packages, being able to create `--canary` builds on only packages that have actually changed since the last release would save significant time.  In my project, this would reduce the build time by at least 80%.

**Comment by bradleat** [comment]
Is this going to be merged in soon?

**Comment by evocateur** [comment]
#963 needs tests, and updating due to changes in #1020

**Comment by bladedancer** [comment]
I'm hitting the exact same issue and started looking at the code to create an up-to-date PR but the more I dig the more something feels wrong here. So before spending too much time on it I wanted to check what's the acceptable behavior for canary.

Our CI system is publishing canary builds on feature branches to essentially create a testable but disposable artifact. We can then independently QA the features prior to merging down to master. The merge to master is publishing the normal build.

Anyway as noted here, the canary publish is publishing all packages. This is noisy, slower and also consumes a lot of space in our private registry (our CI system is doing this on every push). So we're looking into limiting this to just publishing deltas.

The --since option seems to be a workable solution but just in general it feels like I'm abusing the --canary option. The docs describe it's usage as being for per commit publishes and not as being for disposable builds.
Specifically code like this in [UpdatedPackagesCollector](https://github.com/lerna/lerna/blob/master/src/UpdatedPackagesCollector.js#L65)
seems to require it's per commit, i.e. that diff will just return the latest commit, if I've multiple commits in a single push then the canary only sees the latest. Which I suppose is in line with what the docs say. 

So long story short, if --canary got a --since flag would it be in line with the expectations of the canary functionality?

thanks.

**Comment by evocateur** [comment]
Does the solution described in https://github.com/lerna/lerna/issues/1161#issuecomment-359626768 get you closer to this sort of thing? It should get you "partial" publishing, and you could even add `--since staging` in the feature branches to get exactly what has changed in the branch.

> `lerna publish --yes --skip-git --exact --cd-version=prerelease --preid=alpha.${COMMIT_SHA:0:6}`

I agree that `--canary --since` smells a bit more like abusing `--canary` than I'm entirely comfortable with. My understanding of `--canary` is that it was generally intended to be a "nightly" build of `master`, not "in-development" branches.

The reason it bumps everything is that (as highlighted in #277) it does not "properly" increment the prerelease tag, thus you can't use semver ranges like `^1.1.0-alpha.0` in your consumers and magically get all future canary builds from that compatible range (e.g., `1.1.0-alpha.1`, `1.1.0-alpha.2`, etc etc). You currently have to use the full explicit versions (`1.1.0-alpha.{SHA}`), and those packages should have explicit dependencies on the nightly stuff as of that commit, _not_ "ahead" or "behind" from public packages.

**Comment by darrylhodgins** [comment]
Perhaps it shouldn't be `--canary` then, but something else entirely?  Our process is to publish the latest changes from a development branch into an integration test environment.  It would be nice if we didn't have to publish packages that haven't actually changed.

**Comment by yoitsro** [comment]
I think this issue is closed with lerna 3 now, right?

However, for some reason, when a canary publish is performed, the packages which depend on the specific canary versions (e.g. `1.10.3-preid.hash.0`) of those packages aren't updated to include these canary versions. Is that expected behaviour, or a bug?

I've tried what's suggested here https://github.com/lerna/lerna/issues/1161#issuecomment-359626768, which worked in lerna 2, but not lerna 3 (even with the `bump` positional). Any ideas?

**Comment by yoitsro** [comment]
So, I've just seen that the versions all get bumped whilst the publish is in progress, but then reset after the publish occurs. This is generally fine since we don't want a dirty working tree, but I do need a way to consume the canary packages from within a docker build.

I've tried to use the `version` command, but it doesn't get the ref count, which is what the canary versions are published with. Any help would be ideal! Thank you!
